server/src/main/java/org/opensearch/index/mapper/BooleanFieldMapper.java

```diff
@@ -257,9 +272,77 @@ public DocValueFormat docValueFormat(@Nullable String format, ZoneId timeZone) {
             return DocValueFormat.BOOLEAN;
         }
 
+        @Override
+        public Query termQuery(Object value, QueryShardContext context) {
+            failIfNotIndexedAndNoDocValues();
+            Query query = new TermQuery(new Term(name(), indexedValueForSearch(value)));
+            if (boost() != 1f) {
+                query = new BoostQuery(query, boost());
+            }
+            if (isSearchable() && hasDocValues()) {
+                return new IndexOrDocValuesQuery(
+                    query,
+                    SortedNumericDocValuesField.newSlowExactQuery(name(), Values.TRUE.bytesEquals(indexedValueForSearch(value)) ? 1 : 0)
+                );
+            }
+            if (hasDocValues()) {
+                return SortedNumericDocValuesField.newSlowExactQuery(name(), Values.TRUE.bytesEquals(indexedValueForSearch(value)) ? 1 : 0);
+            }
+            return query;
+        }
+
+        @Override
+        public Query termsQuery(List<?> values, QueryShardContext context) {
+            failIfNotIndexedAndNoDocValues();
+            boolean seenTrue = false;
+            boolean seenFalse = false;
+            for (Object value : values) {
+                if (Values.TRUE.equals(indexedValueForSearch(value))) seenTrue = true;
+                else seenFalse = true;
+            }
+            if (seenTrue) {
+                if (seenFalse) {
+                    return new MatchAllDocsQuery();
```

Comment:
I don't think this would be a correct substitution - what if document has no field in question at all?

Reply:
IIRC I think we handle the field level check earlier in the search query? Don't we evaluate the field exists before the mapper can call the `termsQuery`? I might have to double check

Reply:
@harshavamsi curious to know what you have found there, do we evaluate the field exists before the mapper can call the `termsQuery`?

Reply:
@reta surprisingly I thought that the `TermsQueryBuilder` would have field level logic to verify that a query on the field during XContent parsing would be validated but I did not see that. I did perform a simple experiment:

Created an index with with my change:

```
{
  "mappings": {
    "properties": {
      "dec": {
        "type": "boolean"
      },
      "undec": {
        "type": "keyword"
      }
    }
  }
}
```


Indexed two docs:

```

{
  "dec": true,
  "undec": "key1"
}
```

```
{
  "undec": "key1"
}
```


Ran this query:

```
{
    "query": {
        "terms": {
            "dec": [true]
        }
    }
}
```

Got this back:

```
{
    "took": 3,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 1,
            "relation": "eq"
        },
        "max_score": 0.2876821,
        "hits": [
            {
                "_index": "new_index",
                "_id": "1",
                "_score": 0.2876821,
                "_source": {
                    "dec": true,
                    "undec": "key1"
                }
            }
        ]
    }
}
```

When I ran it with this

```
{
    "query": {
        "terms": {
            "dec": [true, false]
        }
    }
}
```

I expected to see the same response as above but I see 

```
{
    "took": 2,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 2,
            "relation": "eq"
        },
        "max_score": 1.0,
        "hits": [
            {
                "_index": "new_index",
                "_id": "1",
                "_score": 1.0,
                "_source": {
                    "dec": true,
                    "undec": "key1"
                }
            },
            {
                "_index": "new_index",
                "_id": "2",
                "_score": 1.0,
                "_source": {
                    "undec": "key2"
                }
            }
        ]
    }
}
```


While testing it on a field that does not exist I get no hits:

```
{
    "query": {
        "terms": {
            "diff": [true]
        }
    }
}
```


```
{
    "took": 1,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 0,
            "relation": "eq"
        },
        "max_score": null,
        "hits": []
    }
}
```

So I think this case is fine, I will rework this function and write some tests.

Reply:
@reta sorry for the delay but finally addressed this. I ended up going to the `TermsInSetQuery` which is the default lucene implementation. If we have only doc_values, I wrap the query in `SortedNumericDocValuesQuery` and do a `newSlowSetQuery` since booleans are internally converted to 0 and 1. Updated unit tests and added rest yaml tests for this change.